### PR TITLE
fix: sealed header should not be immutable borrowed

### PR DESCRIPTION
**Pull request by yjhmelody**
- Fix `pending_block_env_and_cfg` logic problem caused by the mutable ops but `sealedHeader` hash is not updated. It seems influence so many eth apis.  

The following improvements are useful to avoid the problem: 
- The `sealed` type should always means it's not easy to change its value directly. So should remove all immutable borrowed methods and replace pub fields by ref apis.
- Add some mutable api to SealedHeader but with `test-utils` feature. In non-test code path, SealedHeader should always be created by Header and then be immutable.

BTW I think other sealed types also need to be encapsulated and refactorred.

### CODE REVIEW

File: crates/primitives/src/header.rs [hunk 1]
```diff
@@ -670,12 +667,66 @@ pub enum HeaderValidationError {
...
+
+    /// Returns the sealed Header fields.
+    #[inline]
+    pub fn header(&self) -> &Header {
+        &self.header
+    }
+
+    /// Returns header/block hash.
+    #[inline]
+    pub fn hash(&self) -> BlockHash {
```
Comment (mattsse) [inline crates/primitives/src/header.rs:690]:
```suggestion
    pub const fn hash(&self) -> BlockHash {
```

File: crates/rpc/rpc/src/eth/api/mod.rs [hunk 2]
```diff
@@ -270,22 +271,29 @@ where
...
             // base fee of the child block
             let chain_spec = self.provider().chain_spec();
-            latest.base_fee_per_gas =
-                latest.next_block_base_fee(chain_spec.base_fee_params(latest.timestamp));
+
+            latest_header.base_fee_per_gas = latest_header
+                .next_block_base_fee(chain_spec.base_fee_params(latest_header.timestamp));
+
+            let block_hash = latest_header.hash_slow();
+            let latest = SealedHeader::new(latest_header, block_hash);
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/mod.rs:296]: nice fix

**Review by mattsse — APPROVED** [review]
oh, using a function for the hash is actually better imo

I think we should do this, ptal @rkrasiuk @Rjected @onbjerg 

pending test failures

**Review by Rjected — APPROVED** [review]
yeah, I like using a function for this, lgtm!

**Review by onbjerg — APPROVED** [review]
lgtm too

**Comment by mattsse** [comment]
sweet, sending it before this introduces a bunch of conflicts

**Pull request merged**